File: Chap8-Internationalization/8-1InternationalizationDemo/commands.py
from PySide6.QtGui import QUndoCommand

class AddCommand(QUndoCommand):
    """Command for adding items to the scene"""
    
    def __init__(self, item, scene):
        super().__init__()
        self.item = item
        self.scene = scene
    
    def undo(self):
        """Remove the item from the scene"""
        if self.item:
            self.scene.removeItem(self.item)
    
    def redo(self):
        """Add the item to the scene"""
        if self.item:
            self.scene.addItem(self.item)

class RemoveCommand(QUndoCommand):
    """Command for removing items from the scene"""
    
    def __init__(self, item, scene):
        super().__init__()
        self.item = item
        self.scene = scene
    
    def undo(self):
        """Add the item back to the scene"""
        if self.item:
            self.scene.addItem(self.item)
    
    def redo(self):
        """Remove the item from the scene"""
        if self.item:
            self.scene.removeItem(self.item)

class MoveCommand(QUndoCommand):
    """Command for moving items in the scene"""
    
    def __init__(self, item, scene, old_pos, new_pos):
        super().__init__()
        self.item = item
        self.scene = scene
        self.old_position = old_pos
        self.new_position = new_pos
    
    def undo(self):
        """Move the item back to its old position"""
        if self.item:
            self.item.setPos(self.old_position)
    
    def redo(self):
        """Move the item to its new position"""
        if self.item:
            self.item.setPos(self.new_position)

File: 18-Internationalization/8-1InternationalizationDemo/commands.cpp
#include "commands.h"

AddCommand::AddCommand(QGraphicsItem * item, QGraphicsScene * scene) :
    mItem(item) , mScene(scene)
{

}

void AddCommand::undo()
{
    if(mItem){
        mScene->removeItem(mItem);
    }
}

void AddCommand::redo()
{
    if(mItem){
        mScene->addItem(mItem);
    }


}

RemoveCommand::RemoveCommand(QGraphicsItem *item, QGraphicsScene *scene):
    mItem(item) , mScene(scene)
{

}

void RemoveCommand::undo()
{
    if(mItem){
        mScene->addItem(mItem);
    }
}

void RemoveCommand::redo()
{
    if(mItem)
        mScene->removeItem(mItem);
}

MoveCommand::MoveCommand(QGraphicsItem *item,
                         QGraphicsScene *scene, QPointF oldPos, QPointF newPos) :
    mItem(item) , mScene(scene), mOldPosition(oldPos),
    mNewPosition(newPos)
{

}

void MoveCommand::undo()
{
    if(mItem){
        mItem->setPos(mOldPosition);
    }

}

void MoveCommand::redo()
{
    if(mItem)
        mItem->setPos(mNewPosition);
}
